pyalgostrategypool/bollinger_bands.py

```diff
@@ -0,0 +1,207 @@
+        """
+
+        super().__init__(*args, **kwargs)
+
+        # Bollinger Bands parameters
+        self.time_period = self.strategy_parameters['TIME_PERIOD']
+        self.std_deviations = self.strategy_parameters['STANDARD_DEVIATIONS']
+
+        # Sanity
+        assert (0 < self.time_period == int(self.time_period)), f"Strategy parameter TIME_PERIOD should be a positive integer. Received: {self.time_period}"
+        assert (0 < self.std_deviations == int(self.std_deviations)), f"Strategy parameter STANDARD_DEVIATIONS should be a positive integer. Received: {self.std_deviations}"
+
+        # Variables
+        self.main_order = None
+
+    def initialize(self):
+        """
+        If you are running the strategy for multiple days, then this method will be called once at the start of every day
+        use this to initialize and re-initialize your variables
+        """
+        self.main_order = {}
+
+    @staticmethod
+    def name():
+        """
+        Name of your strategy
+        """
+        return 'Bollinger Bands'
+
+    @staticmethod
+    def versions_supported():
+        """
+        Strategy should always support the latest engine version
+        Current version is 3.3.0
+        """
+        return [AlgoBullsEngineVersion.VERSION_3_3_0]
+
+    def get_decision(self, instrument):
+        """
+        This method calculates the crossover using the hist data of the instrument along with the required indicator and returns the entry/exit action
+        """
+
+        # Get OHLC historical data for the instrument
+        hist_data = self.get_historical_data(instrument)
+
+        # Get last OHLC row of the historical data
+        latest_candle = hist_data.iloc[-1]
+
+        # Get second last OHLC row of the historical data
+        previous_candle = hist_data.iloc[-2]
+
+        # Calculate the Bollinger Bands values
+        upperband, _, lowerband = talib.BBANDS(hist_data['close'], timeperiod=self.time_period, nbdevup=self.std_deviations, nbdevdn=self.std_deviations, matype=0)
+        upperband_value = upperband.iloc[-1]
+        lowerband_value = lowerband.iloc[-1]
+
+        # Check entry/exit conditions
+        if (previous_candle['open'] <= lowerband_value or previous_candle['high'] <= lowerband_value or previous_candle['low'] <= lowerband_value or previous_candle['close'] <= lowerband_value) and \
+                (latest_candle['close'] > previous_candle['close']):
+            action = 'BUY'
+        elif (previous_candle['open'] >= upperband_value or previous_candle['high'] >= upperband_value or previous_candle['low'] >= upperband_value or previous_candle['close'] >= upperband_value) and \
+                (latest_candle['close'] < previous_candle['close']):
+            action = 'SELL'
+        else:
+            action = None
+        logger.info(f'Previous candle: {[previous_candle]}\n'
```

Comment:
why do we have [ ] for previous candle ?

dont print the complete candle

print ohlc of previous candle in one line, then `\n`, then latest close upper band and lower band in second line, then `\n`, then action in the third line

Reply:
Done